neural_compressor/adaptor/torch_utils/teq.py

```diff
@@ -0,0 +1,485 @@
+
+    @torch.no_grad()
+    def transform(self):
+        """
+        apply alpha/scale
+        """
+        for ln_name, layer_names in self.absorb_to_layer.items():
+            module = get_module(self.model, ln_name)
+            scale = self.trained_alphas[ln_name]
+            scale = torch.clip(scale, 1e-5)
+            input_scale = 1.0 / scale
+            if hasattr(module, "orig_layer"):
+                module = module.orig_layer
+
+            self._absorb_scales(module, input_scale, layer_name=ln_name)
+            weight_scale = scale
+            for layer_name in layer_names:
+                layer_module = get_module(self.model, layer_name)
+                self._scale_layer_weight(layer_module, weight_scale)
+
+        # for insert_mul = False
+        for n, m in self.model.named_modules():
+            if isinstance(m, WrapperLinear):
+                set_module(self.model, n, m.orig_layer)
+
+    def _trace(self, op_types):
+        """
+        Try the model to find the layers which can be smooth quantized.
+        :param op_types: The op types to be smooth quantized
+        :return:
+        absorb_to_layer: A dict, absorb layer name:layers to be smooth quantized
+        no_absorb_layers: A list saving the layers which could not find the absorb layer
+        """
+        tg = GraphTrace()
+        self._get_example_input()
+        absorb_to_layer, no_absorb_layers = tg.get_absorb_to_layer(self.traced_model, self.example_inputs, op_types)
+        return absorb_to_layer, no_absorb_layers
+
+    def train(self, train_steps=100, lr=1e-3, warmup_ratio=0.05, gradient_accumulation_steps=1, logging_steps=10,
+            betas=[0.9, 0.9], weight_decay=0, optimizer=None, lr_scheduler=None, lr_scheduler_type="linear"):
+        """
+        train function
+        """
+        if optimizer is None:
+            trained_alphas_list = []
+            for item in self.trained_alphas.items():
+                trained_alphas_list.append(item[1])
+            optimizer = torch.optim.Adam(trained_alphas_list, lr=lr, weight_decay=weight_decay, betas=betas)
+
+
+        if lr_scheduler is None:
+            lr_scheduler = get_scheduler(
+                    name=lr_scheduler_type,
+                    optimizer=optimizer,
+                    num_warmup_steps=int(train_steps * warmup_ratio) // gradient_accumulation_steps,
+                    num_training_steps=train_steps // gradient_accumulation_steps,
+                    )
+
+        logger.info("start training")
+        self.model.train()
+        global_steps = 0
+
+        while True:
+            for inputs in self.dataloader:
+                if isinstance(inputs, dict):
+                    input_id = inputs["input_ids"]
+                else:
+                    input_id = inputs[0]
+
+                input_id = input_id.to(self.device)
+                output = self.model(input_id, labels=input_id)
+                loss = output[0] / gradient_accumulation_steps
```

Comment:
need to support customized loss function later

Reply:
train func will be moved outside for custom use.

Reply:
better have a default one in the code